### PR TITLE
40 rewrite the ml model

### PR DESCRIPTION
**Pull request by Varlog22**
Made our first full ML-model for the QAQC dataset

### CODE REVIEW

File: Backend/src/ml_lib/shared/shared_model.py [hunk 1]
```diff
@@ -22,7 +22,7 @@ def _format_data(self, file_name: str) -> None:
         """
         pass
 
-    def run(self, file_name: str, is_first: bool) -> None:
+    def run(self, file_name: str, is_first: bool, dim: str = None) -> None:
```
Comment (Leohemmingsson) [inline Backend/src/ml_lib/shared/shared_model.py:25]: I think that the type for "dim" should be `str | None` . This also applies in the sub_models

File: Backend/src/v1/run_ml.py [hunk 2]
```diff
@@ -11,26 +11,29 @@
     tags=["Run ML"],
 )
 
-
 @router.get("/", status_code=200)
-def run(model_name: str, file: str) -> dict:
+def run(model_name: str, file: str, dim_red_method: str) -> dict:
```
Comment (Leohemmingsson) [inline Backend/src/v1/run_ml.py:15]: If you set `dim_red_method: str | None = None`  you can run without giving a value, which I think makes more sense especially if running the test_model. 

If all methods **needs** to have a dim_red_method, I think it might be a good thing to specify a default value here instead of using `=None`. This would also remove the need to fill with a  dummy value when running the test model.

File: Backend/src/ml_lib/sub_models/qaqc_main_model.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        return re.sub(pattern, replace_commas, line)
+
+    def first_run(self, df: pd.DataFrame, dim: str) -> pd.DataFrame:
+        """
+        Combines the input question data with the calculated 2D point data
+        """
+        questions = df["text"].tolist()
+        model = SentenceTransformer("paraphrase-MiniLM-L6-v2")
+        embeddings = model.encode(questions, convert_to_tensor=True)
+        match dim:
```
Comment (Leohemmingsson) [inline Backend/src/ml_lib/sub_models/qaqc_main_model.py:69]: I think it could be a good thing to add a default case which would raise a exception that the dim does not exist.

I belive that the code for a default case is `case _:`

File: Backend/src/ml_lib/sub_models/qaqc_main_model.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            case "COMBO":
+                pca2 = PCA(n_components=50, whiten=False, random_state=42)
+                reduced_embeddings_pca2 = pca2.fit_transform(embeddings)
+                tsne = TSNE(n_components=2, random_state=42, perplexity=30, n_iter=300, early_exaggeration=4, learning_rate=1000)
+                reduced_embeddings = tsne.fit_transform(reduced_embeddings_pca2)
+                point_data_df_COMBO = pd.DataFrame(reduced_embeddings, columns=["x", "y"])
```
Comment (Leohemmingsson) [inline Backend/src/ml_lib/sub_models/qaqc_main_model.py:75]: I think it could make the code a lite less cluttered if all the cases created a pd Dataframe e.g. `point_data_df`, and then after the match statement combine the original df with the points.

Ex)
```python
case: "PCA":
  ...
  point_data_df = ...
  break
case: "TSNE":
  ...
  point_data_df = ...
  break

combined_df= pd.concat([df, point_data_df, axis=1)
combined_df = combined_df.rename(columns={"coarse label": "truth"})
return combined_df
```

File: Backend/src/ml_lib/sub_models/qaqc_main_model.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+from sentence_transformers import SentenceTransformer
+from sklearn.decomposition import PCA
+from sklearn.manifold import TSNE
+
+from ..shared import IsbitClassifierModel
+
+
+class QaqcMainModel(IsbitClassifierModel):
+
+    def _format_data(self, file_name: str) -> None:
```
Comment (Leohemmingsson) [inline Backend/src/ml_lib/sub_models/qaqc_main_model.py:15]: I saw that the `_format_data` looks identical to the test model version, do we know if  it will be used for a lot of different models/all the models?

If so, could we break out the functionality into a separate file? Making less redundant code?

**Review by Leohemmingsson — CHANGES_REQUESTED** [review]
Good job! 

When running the model it looks good, but I think there is a bug somewhere. Not quite sure why it happened, I only got it once. But I am documenting it here for traceability/if it is something anyone recognizes.

I also wrote some comments of things I think could improve the code a little. 

```
backend-1   | Error while downloading from https://cdn-lfs.hf.co/sentence-transformers/paraphrase-MiniLM-L6-v2/2ce4480dc3b2f8edeee50c43765c72768e79fc0113d3f73773dded4887cca298?response-content-disposition=inline%3B+filename*%3DUTF-8%27%27model.safetensors%3B+filename%3D%22model.safetensors%22%3B&Expires=1730315709&Policy=eyJTdGF0ZW1lbnQiOlt7IkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTczMDMxNTcwOX19LCJSZXNvdXJjZSI6Imh0dHBzOi8vY2RuLWxmcy5oZi5jby9zZW50ZW5jZS10cmFuc2Zvcm1lcnMvcGFyYXBocmFzZS1NaW5pTE0tTDYtdjIvMmNlNDQ4MGRjM2IyZjhlZGVlZTUwYzQzNzY1YzcyNzY4ZTc5ZmMwMTEzZDNmNzM3NzNkZGVkNDg4N2NjYTI5OD9yZXNwb25zZS1jb250ZW50LWRpc3Bvc2l0aW9uPSoifV19&Signature=KvkQ0--x-Zz0c0SzbhO5IZDa9IuubtAd%7EB7uGm7bVKn2%7EyXBmso04rc4DYcCYblyoy5pfjyiRv42ZpiKJoS8kOUqUBCwA7mGlf0-vNmppnwmLuZsoNXaSPP-vJAibok11AvVoWmkl7ZNbI39G1UjggSHI0p0yMC3fgdQwi9G%7Ej3gHtgOEoOIQ4wck1aiJ5WQ3we2FCaQA0O9xy7CGtyt12w5SdMvH-joW7RVq1vVQgPRcIhdZLVxHrDNY-DqhfImYjvqCJTmt-hp5WkU15ThNC3Wm-IIPghDuz4x1hVWXEZ9q2iHpvcZ9MmoioYwweVW-a9ynAF-s2iVZWIeBb2M8Q__&Key-Pair-Id=K3RPWS32NSSJCE: HTTPSConnectionPool(host='cdn-lfs.hf.co', port=443): Read timed out.
```

File: Backend/src/v1/run_ml.py (hunk 2 above)
Comment (Varlog22) [reply Backend/src/v1/run_ml.py:15]: Perfect, I was not sure about how to implement that, thanks for the tip

File: Backend/src/ml_lib/sub_models/qaqc_main_model.py (hunk 3 above)
Comment (Varlog22) [reply Backend/src/ml_lib/sub_models/qaqc_main_model.py:69]: done

File: Backend/src/ml_lib/sub_models/qaqc_main_model.py (hunk 4 above)
Comment (Varlog22) [reply Backend/src/ml_lib/sub_models/qaqc_main_model.py:75]: great idea, implemented it as well.

File: Backend/src/ml_lib/sub_models/qaqc_main_model.py (hunk 5 above)
Comment (Leohemmingsson) [reply Backend/src/ml_lib/sub_models/qaqc_main_model.py:15]: Sounds good

File: Backend/src/ml_lib/shared/shared_model.py [hunk 6]
```diff
@@ -32,11 +32,11 @@ def run(self, file_name: str, is_first: bool) -> None:
         if not is_first:
             df = self.latter_run(data)
         else:
-            df = self.first_run(data)
+            df = self.first_run(data, dim)
 
         self._df_setter(df)
 
-    def first_run(self, df: pd.DataFrame) -> pd.DataFrame:
+    def first_run(self, df: pd.DataFrame, dim: str = None) -> pd.DataFrame:
```
Comment (Leohemmingsson) [inline Backend/src/ml_lib/shared/shared_model.py:39]: The type for dim should be `str | None` here as well, sorry I missed this on the last review. Same should apply to the sub_models

**Review by Leohemmingsson — CHANGES_REQUESTED** [review]
The changes looks great, found one last thing

File: Backend/src/ml_lib/shared/shared_model.py (hunk 6 above)
Comment (Varlog22) [reply Backend/src/ml_lib/shared/shared_model.py]: done

**Review by Leohemmingsson — APPROVED** [review]
Looks all good!

**Review by SpiffiRacoon — APPROVED** [review]
lgtm 🦝
